yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts

```diff
@@ -300,4 +342,181 @@ export class AztecKVTxPool implements TxPool {
+   * @returns The total number of txs evicted from the pool and the number of new txs that were evicted.
+   *
+   * TODO: add logging for evicted txs?
+   */
+  private async evictLowPriorityTxs(
+    newTxHashes: TxHash[],
+  ): Promise<{ numLowPriorityTxsEvicted: number; numNewTxsEvicted: number }> {
+    let pendingTxsByPriority = (await toArray(this.#pendingTxPriorityToHash.valuesAsync())).map(TxHash.fromString);
+    const pendingTxCount = pendingTxsByPriority.length;
+    let numNewTxsEvicted = 0;
+    const txsToEvict: TxHash[] = [];
+
+    if (pendingTxCount > this.#maxTxPoolCount) {
+      const numTxsToEvict = pendingTxCount - this.#maxTxPoolCount;
+      txsToEvict.push(...pendingTxsByPriority.slice(0, numTxsToEvict));
+      numNewTxsEvicted += newTxHashes.filter(txHash => txsToEvict.includes(txHash)).length;
+
+      pendingTxsByPriority = pendingTxsByPriority.slice(numTxsToEvict);
+    }
+
+    let pendingTxsSize = (await this.#pendingTxSize.getAsync()) ?? 0;
+    if (pendingTxsSize > this.#maxTxPoolSize) {
+      for (const txHash of pendingTxsByPriority) {
+        txsToEvict.push(txHash);
+
+        const txSize = (await this.#pendingTxHashToSize.getAsync(txHash.toString())) ?? 0;
+        pendingTxsSize -= txSize;
+        if (pendingTxsSize <= this.#maxTxPoolSize) {
+          break;
+        }
+      }
+      numNewTxsEvicted += newTxHashes.filter(txHash => txsToEvict.includes(txHash)).length;
+    }
+
+    if (txsToEvict.length > 0) {
+      await this.deleteTxs(txsToEvict, true);
+    }
+    return {
+      numLowPriorityTxsEvicted: txsToEvict.length,
+      numNewTxsEvicted,
+    };
+  }
+
+  /**
+   * Evicts invalid pending txs from the pool after a txs from a block are mined.
+   * Eviction criteria includes:
+   *   - txs with nullifiers that are already included in the mined block
+   *   - txs with an insufficient fee payer balance
+   *   - txs with a max block number lower than the mined block
+   *
+   * @param minedTxHashes - The tx hashes of the txs mined in the block.
+   * @param blockNumber - The block number of the mined block.
+   * @returns The total number of txs evicted from the pool.
+   *
+   * TODO: add logging for evicted txs?
+   */
+  private async evictInvalidTxsAfterMining(minedTxHashes: TxHash[], blockNumber: number): Promise<number> {
+    if (minedTxHashes.length === 0) {
+      return 0;
+    }
+
+    const pendingTxHashes = await this.getPendingTxHashes();
+    const minedTxs = await Promise.all(minedTxHashes.map(async hash => await this.getTxByHash(hash)));
+    const minedNullifiers = new Set<Fr>();
+    for (const tx of minedTxs) {
+      if (tx) {
+        const nullifiers = tx.data.getNonEmptyNullifiers();
+        nullifiers.forEach(nullifier => minedNullifiers.add(nullifier));
+      }
+    }
+    const pendingTxs = await Promise.all(
+      pendingTxHashes.map(async hash => ({
+        hash,
+        tx: await this.getTxByHash(hash),
```

Comment:
Once an initial version of this PR is done, let's benchmark how expensive it is to rehydrate all txs, since we'll be deserializing all txs in the mempool every 36s. It may be worthwhile keeping the hydrated instances in memory for easy access, if it's not too much memory and saves time.